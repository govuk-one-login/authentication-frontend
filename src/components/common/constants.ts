import { PATH_NAMES } from "../../app.constants";
import { getNextState } from "./state-machine/state-machine";

export const SECURITY_CODE_ERROR = "actionType";

export enum SecurityCodeErrorType {
  MfaMaxCodesSent = "mfaMaxCodesSent",
  MfaBlocked = "mfaBlocked",
  MfaMaxRetries = "mfaMaxRetries",
  OtpMaxCodesSent = "otpMaxCodesSent",
  OtpBlocked = "otpBlocked",
  OtpMaxRetries = "otpMaxRetries",
  EmailMaxCodesSent = "emailMaxCodesSent",
  EmailBlocked = "emailBlocked",
  EmailMaxRetries = "emailMaxRetries",
  AuthAppMfaNoCodeValidator = "authAppMfaNoCodeValidator",
  AuthAppMfaMaxRetries = "authAppMfaMaxRetries",
  ChangeSecurityCodesEmailMaxCodesSent = "changeSecurityCodesEmailMaxCodesSent",
  ChangeSecurityCodesEmailBlocked = "changeSecurityCodesEmailBlocked",
  ChangeSecurityCodesEmailMaxRetries = "changeSecurityCodesEmailMaxRetries",
  InvalidPasswordResetCodeMaxRetries = "invalidPasswordResetCodeMaxRetries",
}

export const ERROR_CODES = {
  SESSION_ID_MISSING_OR_INVALID: 1000,
  RESET_PASSWORD_LINK_MAX_RETRIES_REACHED: 1022,
  RESET_PASSWORD_LINK_BLOCKED: 1023,
  NEW_PASSWORD_SAME_AS_EXISTING: 1024,
  ENTERED_INVALID_MFA_MAX_TIMES: 1027,
  VERIFY_EMAIL_CODE_REQUEST_BLOCKED: 1029,
  VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED: 1030,
  VERIFY_PHONE_NUMBER_MAX_CODES_SENT: 1032,
  VERIFY_EMAIL_MAX_CODES_SENT: 1031,
  ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES: 1033,
  ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES: 1034,
  MFA_SMS_MAX_CODES_SENT: 1025,
  MFA_CODE_REQUESTS_BLOCKED: 1026,
  INVALID_MFA_CODE: 1035,
  INVALID_VERIFY_EMAIL_CODE: 1036,
  INVALID_VERIFY_PHONE_NUMBER_CODE: 1037,
  ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES: 1039,
  INVALID_PASSWORD_MAX_ATTEMPTS_REACHED: 1028,
  RESET_PASSWORD_INVALID_CODE: 1021,
  PASSWORD_IS_COMMON: 1040,
  AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED: 1042,
  AUTH_APP_INVALID_CODE: 1043,
  ACCOUNT_LOCKED: 1045,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT: 1046,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED: 1047,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE: 1048,
};

export const ERROR_CODE_MAPPING: { [p: string]: string } = {
  [ERROR_CODES.ACCOUNT_LOCKED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.MFA_SMS_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxCodesSent
  ),
  [ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpMaxCodesSent
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.OtpMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxRetries
  ),
  [ERROR_CODES.ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.InvalidPasswordResetCodeMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxCodesSent
  ),
  [ERROR_CODES.AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.AuthAppMfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxCodesSent
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxRetries
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_WAIT"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailBlocked
    ),
};

export function getErrorPathByCode(errorCode: number): string | undefined {
  const nextPath = ERROR_CODE_MAPPING[errorCode.toString()];

  if (!nextPath) {
    return undefined;
  }

  return nextPath;
}

export function pathWithQueryParam(
  path: string,
  queryParam?: string,
  value?: string | SecurityCodeErrorType
): string {
  if (queryParam && value) {
    const queryParams = new URLSearchParams({
      [queryParam]: value,
    }).toString();

    return path + "?" + queryParams;
  }

  return path;
}

export function getNextPathAndUpdateJourney(
  req: any,
  path: string,
  event: string,
  ctx?: any,
  sessionId?: string
): string {
  const nextState = getNextState(path, event, ctx);

  req.session.user.journey = {
    nextPath: nextState.value,
    optionalPaths:
      Object.keys(nextState.meta).length > 0
        ? nextState.meta["AUTH." + nextState.value].optionalPaths
        : [],
  };

  req.log.info(
    `User journey transitioned from ${req.path} to ${nextState.value} with session id ${sessionId}`
  );

  if (!nextState) {
    throw Error(
      `Invalid user journey. No transition found from ${path} with event ${event}`
    );
  }

  return nextState.value;
}

export const JOURNEY_TYPE = {
  SIGN_IN: "sign-in",
  CREATE_ACCOUNT: "create-account",
};
