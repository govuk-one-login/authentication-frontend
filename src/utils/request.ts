import type { Request } from "express";
import { supportAccountRecovery, supportReauthentication } from "../config.js";
import {
  CONTACT_US_THEMES,
  SERVICE_TYPE,
  SUPPORT_TYPE,
} from "../app.constants.js";
export const isReauth = (req: Request): boolean =>
  supportReauthentication() && Boolean(req.session?.user?.reauthenticate);

export const isUpliftRequired = (req: Request): boolean =>
  Boolean(req.session?.user?.isUpliftRequired);

export const isAccountRecoveryJourney = (req: Request): boolean =>
  Boolean(req.session?.user?.isAccountRecoveryJourney);

export const isAccountRecoveryJourneyAndEnabled = (req: Request): boolean =>
  Boolean(
    supportAccountRecovery() &&
      req.session?.user?.isAccountRecoveryJourney &&
      req.session?.user?.isAccountRecoveryPermitted
  );

export const isContactUsSuggestionsFeedbackSubtheme = (req: Request): boolean =>
  req.query?.subtheme === CONTACT_US_THEMES.SUGGESTIONS_FEEDBACK;

export const clientIsOneLogin = (req: Request): boolean =>
  Boolean(req.session?.client?.isOneLoginService);

export const clientUsesOneLoginOptionally = (req: Request): boolean =>
  req.session?.client?.serviceType === SERVICE_TYPE.OPTIONAL;

export const supportTypeIsGovService = (req: Request): boolean =>
  Boolean(
    req.query?.supportType === SUPPORT_TYPE.GOV_SERVICE ||
      req.body?.supportType === SUPPORT_TYPE.GOV_SERVICE
  );

export const urlContains = (req: Request, str: string): boolean =>
  Boolean(req.originalUrl?.includes(str));
