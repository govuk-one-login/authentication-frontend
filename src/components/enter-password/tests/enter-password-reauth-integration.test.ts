import { describe } from "mocha";
import { expect, sinon } from "../../../../test/utils/test-utils.js";
import nock from "nock";
import request from "supertest";
import * as cheerio from "cheerio";
import { API_ENDPOINTS, PATH_NAMES } from "../../../app.constants.js";
import { ERROR_CODES } from "../../common/constants.js";
import type { NextFunction, Request, Response } from "express";
import { getPermittedJourneyForPath } from "../../../../test/helpers/session-helper.js";
import esmock from "esmock";
import { buildMfaMethods } from "../../../../test/helpers/mfa-helper.js";

const REDIRECT_URI = "https://rp.host/redirect";

describe("Integration::enter password", () => {
  let token: string | string[];
  let cookies: string;
  let app: any;
  let baseApi: string;

  const ENDPOINT = "/enter-password";

  before(async () => {
    const { createApp } = await esmock(
      "../../../app.js",
      {},
      {
        "../../../middleware/session-middleware.js": {
          validateSessionMiddleware: sinon.fake(function (
            req: Request,
            res: Response,
            next: NextFunction
          ): void {
            res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";
            res.locals.clientSessionId = "gdsfsfdsgsdgsd-mjdzU854";
            res.locals.persistentSessionId = "dips-123456-abc";

            req.session.user = {
              email: "test@test.com",
              journey: getPermittedJourneyForPath(PATH_NAMES.ENTER_PASSWORD),
              reauthenticate: "subject",
            };

            req.session.client = {
              redirectUri: REDIRECT_URI,
            };

            next();
          }),
        },
      }
    );

    app = await createApp();

    baseApi = process.env.FRONTEND_API_BASE_URL;
    process.env.SUPPORT_REAUTHENTICATION = "1";

    await request(app)
      .get(ENDPOINT)
      .then((res) => {
        const $ = cheerio.load(res.text);
        token = $("[name=_csrf]").val();
        cookies = res.headers["set-cookie"];
      });
  });

  after(() => {
    sinon.restore();
    app = undefined;
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  it("should return enter password page with reauth analytics properties", async () => {
    await request(app).get(ENDPOINT).expect(200);
  });

  it("should return error when csrf not present", async () => {
    await request(app)
      .post(ENDPOINT)
      .type("form")
      .send({
        password: "password",
      })
      .expect(403);
  });

  it("should return validation error when password not entered", async () => {
    await request(app)
      .post(ENDPOINT)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        password: "",
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#password-error").text()).to.contains("Enter your password");
      })
      .expect(400);
  });

  it("should return validation error when password is incorrect", async () => {
    nock(baseApi).post(API_ENDPOINTS.LOG_IN_USER).once().reply(401);

    await request(app)
      .post(ENDPOINT)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        password: "pasasd",
      })
      .expect(function (res) {
        const $ = cheerio.load(res.text);
        expect($("#password-error").text()).to.contains(
          "The password you entered is not correct"
        );
      })
      .expect(400);
  });

  it("should redirect to /auth-code when password is correct (VTR Cm)", async () => {
    nock(baseApi)
      .post(API_ENDPOINTS.LOG_IN_USER)
      .once()
      .reply(200, {
        mfaMethods: buildMfaMethods({
          id: "9b1deb4d-3b7d-4bad-9bdd-2b0d7a3a03d7",
          phoneNumber: "07123456789",
          redactedPhoneNumber: "789",
        }),
      });

    await request(app)
      .post(ENDPOINT)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        password: "password",
      })
      .expect("Location", PATH_NAMES.AUTH_CODE)
      .expect(302);
  });

  it("should sign out of re-authentication flow when incorrect password entered 5 times", async () => {
    nock(baseApi).post(API_ENDPOINTS.LOG_IN_USER).times(6).reply(400, {
      code: ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED,
    });

    await request(app)
      .post(ENDPOINT)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        password: "password",
      })
      .expect("Location", REDIRECT_URI.concat("?error=login_required"))
      .expect(302);
  });

  it("should sign out of re-authentication flow when user has reached limit on other reauth credentials", async () => {
    nock(baseApi).post(API_ENDPOINTS.LOG_IN_USER).times(6).reply(400, {
      code: ERROR_CODES.RE_AUTH_SIGN_IN_DETAILS_ENTERED_EXCEEDED,
    });

    await request(app)
      .post(ENDPOINT)
      .type("form")
      .set("Cookie", cookies)
      .send({
        _csrf: token,
        password: "password",
      })
      .expect("Location", REDIRECT_URI.concat("?error=login_required"))
      .expect(302);
  });
});
