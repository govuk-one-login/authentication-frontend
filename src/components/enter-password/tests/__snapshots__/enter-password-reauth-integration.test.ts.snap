// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration::enter password should return enter password page with reauth analytics properties 1`] = `
Object {
  "contentId": "c6f4fed1-ee6d-4d23-a14f-4466e9c1349c",
  "taxonomyLevel1": "accounts",
  "taxonomyLevel2": "re auth",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter password should return error when csrf not present 1`] = `
Object {
  "contentId": "",
  "taxonomyLevel1": "",
  "taxonomyLevel2": "",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter password should return validation error when password is incorrect 1`] = `
Object {
  "contentId": "c6f4fed1-ee6d-4d23-a14f-4466e9c1349c",
  "taxonomyLevel1": "accounts",
  "taxonomyLevel2": "re auth",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;

exports[`Integration::enter password should return validation error when password not entered 1`] = `
Object {
  "contentId": "c6f4fed1-ee6d-4d23-a14f-4466e9c1349c",
  "taxonomyLevel1": "accounts",
  "taxonomyLevel2": "re auth",
  "taxonomyLevel3": "",
  "taxonomyLevel4": "",
  "taxonomyLevel5": "",
}
`;
